### ISSUE TITLE
Segments and Points cannot be added in Typescript Environment

### ISSUE THREAD

**Issue by tscz** [bbc/peaks.js#281]
I stumbled upon issues programmatically adding segments/points in Typescript. I was not able to compile the demo (https://github.com/bbc/peaks.js/blob/master/demo/webaudio.html) in an React/TypeScript App in Visual Studio Code.

## Expected behavior
The add() function of the Segments API (https://github.com/bbc/peaks.js#instancesegmentsaddsegment) and Points API (https://github.com/bbc/peaks.js#instancepointsaddpoint) can be used in a Typescript Environment as in the demo

```
document!
        .querySelector('button[data-action="add-segment"]')!
        .addEventListener("click", function() {
          peaksInstance!.segments.add({
            startTime: peaksInstance!.player!.getCurrentTime(),
            endTime: peaksInstance!.player!.getCurrentTime() + 10,
            labelText: "Test segment",
            editable: true
          });
        });
```

## Current behavior
The Typescript compiler does not compile the aforementioned code with following errors for Points

`(property) Point.editable?: boolean | undefined
Argument of type '{ time: number; labelText: string; editable: true; }' is not assignable to parameter of type 'Point | Point[]'.
  Property 'update' is missing in type '{ time: number; labelText: string; editable: true; }' but required in type 'Point'.ts(2345)
peaks.js.d.ts(39, 5): 'update' is declared here.`

and for Segments respectively
`Argument of type '{ startTime: number; endTime: number; labelText: string; editable: true; }' is not assignable to parameter of type 'Segment | Segment[]'.
  Property 'update' is missing in type '{ startTime: number; endTime: number; labelText: string; editable: true; }' but required in type 'Segment'.ts(2345)
peaks.js.d.ts(22, 5): 'update' is declared here.`

## Issue
The update function is declared as a non-optional property in the Segment and Point Interface definition in peaks.js.d.ts.

## Possible Solution
Spontaneously, I could not see an obvious solution. Labeling the update function property as optional within the interfaces (if API is intended to be open for overrides) is IMHO not a nice solution because then triggering the update() function is not directly possible in a Typescript environment: The update() function might be undefined and thus the compiler needs a " if (segment.update() !== undefined)" before invocation, i.e.
```
(property) Segment.update?: ((options: Peaks.SegmentUpdateOptions) => void) | undefined
Cannot invoke an object which is possibly 'undefined'.ts(2722)
```

## Related issue https://github.com/bbc/peaks.js/pull/260

**Comment by tscz** [comment]
Hi @chrisn,

as a workaround I surrounded the code parts adding segments/points as in
```
// @ts-ignore-start
    peaksInstance!.segments.add({
    startTime: peaksInstance!.player!.getCurrentTime(),
    endTime: peaksInstance!.player!.getCurrentTime() + 10,
    labelText: "Test segment",
    editable: true
     });
// @ts-ignore-end
```
Have I missed anything or am I using the API incorrectly? 

I really like the peaks.js project and it is a great support for me but I would love to use it within Typescript:-) Can I be of any further assistence regarding this issue (if it is really one) (i.e. example code)?

thanks a lot and regards,
Teje

**Comment by chrisn** [comment]
Sorry, I've not had time to look into this yet. Is this an issue with the `points|segments.add()` or `points|segments.update()` methods, or both?

I'm also still fairly new to TypeScript, so any advice you have on how to fix our [Typescript definitions](https://github.com/bbc/peaks.js/blob/master/peaks.js.d.ts) would be really welcome!

**Comment by tscz** [comment]
Same here, I am also very new to TypeScript and the JavaScript world in general, so please also question my analysis. That's what I found out so far:

The problem is a with PeaksInstance.points./PeaksInstance.segments.add() method caused by the Point/Segment.update() methods, actually. I take the Point API as an example, the Segment API is equivalent.

The Point Interface is 
```TS
  interface Point {
    time: number;
    editable?: boolean;
    color?: string;
    labelText?: string;
    id?: string;
    update: (options: PointUpdateOptions) => void;
  }
```
The method signature for the add() method is
```TS
points: {
      add: (points: Point | Point[]) => void;
```
So Typescript expects a complete Point object or array with declared update() function pointer in its body as parameter if you use the add() function. But this is not what is wanted, there is already a default implementation of the update() function within each Point and it should be available for a caller but should not be a parameter in the constructor of a Point. The implementation in plain Javascript perfectly works fine but the TypeScript compiler does not allow it.

Afaik there is no concept of a default implementation (like in Java) in TypeScript and indeed the compiler is complaining if I patch the typescript definitions and mark it as optional with a questionmark:
```TS
  interface Point {
    ...
    update?: (options: PointUpdateOptions) => void;
  }
```
I can now use the add() function in Typescript but I cannot use the update() function anymore without explicitely checking that it is defined, see
```TSX
          let point: Peaks.Point = {
            time: peaksInstance!.player.getCurrentTime(),
            labelText: "Test point",
            editable: true
          };

          peaksInstance!.points.add(point); // works now
          
          if (point.update !== undefined) point.update({ time: 1 }); // only compiles with the if clause present
```

Long story, short: The update() function must somehow be labeled as a default implementation without forcing the user of the add()-API to define or check it.

**Comment by tscz** [comment]
I can see that on the Typescript Website they use mixins in such cases, see
https://www.typescriptlang.org/docs/handbook/mixins.html. They use a class Activatable as example (in the Point/Segment case it would be more a Updatable class) to model default implementations.

I need to build up more knowledge and have a deeper look to see how to fix the Typescript definition. I will add my findings here.

**Comment by tscz** [comment]
The mixin strategy mentioned above is a dead end street, since it causes production code changes.

One way to fix the Typescript definition would be a similar approach as already done in the SegmentUpdateOptions/PointUpdateOptions: Introduce SegmentAddOptions/PointAddOptions and update the add() method signature as in
```TS
  interface PointAddOptions {
    time: number;
    editable?: boolean;
    color?: string;
    labelText?: string;
    id?: string;
  }

  interface Point extends PointAddOptions {
    update: (options: PointUpdateOptions) => void;
  }

  interface SegmentAddOptions {
    startTime: number;
    endTime: number;
    editable?: boolean;
    color?: string;
    labelText?: string;
    id?: string;
  }

  interface Segment extends SegmentAddOptions {
    update: (options: SegmentUpdateOptions) => void;
  }

    // Segments API
    segments: {
      add: (segments: SegmentAddOptions | SegmentAddOptions[]) => void;
      getSegments: () => Segment[];
      ...
    };
    // Points API
    points: {
      add: (points: PointAddOptions | PointAddOptions[]) => void;
      getPoints: () => Point[];
      ...
    };
```

**Comment by chrisn** [comment]
You're right, I'll add this and publish a new release. Many thanks for figuring this out!

**Comment by chrisn** [comment]
I just published a new release, v0.13.1 with your fix. Please let me know if this works for you.

**Comment by tscz** [comment]
I upgraded to v0.13.1 and now it works for me:-) Thanks a lot for your super quick response and directly releasing it, great!
